### PR TITLE
Transform urldecode 2689 v8

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/2689

Describe changes:
- Adds url_decode transform keyword (and even updates the doc)

There are some open questions :
- should this PR limit the size of the input buffer ? as mentioned here https://redmine.openinfosecfoundation.org/issues/2689#note-1
- What should be the behavior in the case the transform fails ? (that is the buffer is not a valid url-encoded one and has patterns such as `%ZZ`) Right, now I keep undecoded data (like https://www.urldecoder.org for instance)
- The use of a transform looked good to me as it can get reused and is only computed if needed, but there is the suggestion to do it like `http_uri`

There is no reuse of libhtp code as libhtp `htp_decode_path_inplace` mixes more context, such as setting `response_status_expected_number` for the transaction...

Makes test https://github.com/OISF/suricata-verify/pull/146 pass

Modifies #5176 with rebase as requested (to rerun with updated S-V PR)

suricata-verify-pr: 278

### CODE REVIEW

File: src/detect-transform-urldecode.c [hunk 1]
```diff
@@ -0,0 +1,166 @@
+/* Copyright (C) 2019 Open Information Security Foundation
```
Comment (victorjulien) [inline src/detect-transform-urldecode.c:1]: 2020

File: src/detect-http-client-body.c [hunk 2]
```diff
@@ -258,6 +258,7 @@ static InspectionBuffer *HttpClientBodyGetDataCallback(DetectEngineThreadCtx *de
     StreamingBufferGetDataAtOffset(body->sb,
             &data, &data_len, offset);
     InspectionBufferSetup(buffer, data, data_len);
+    InspectionBufferApplyTransforms(buffer, transforms);
```
Comment (victorjulien) [inline src/detect-http-client-body.c:261]: I think https://github.com/OISF/suricata/pull/5213/commits/39760336c22f043c060ac36d64d3dc04dd467e9f should be dropped for now. @jlucovsky is working on addressing to issues around applying transforms to HTTP.

File: src/detect-transform-urldecode.c [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+#include "detect.h"
+#include "detect-engine.h"
+#include "detect-engine-prefilter.h"
+#include "detect-parse.h"
+#include "detect-transform-urldecode.h"
+
+#include "util-unittest.h"
+#include "util-print.h"
+//isxdigit
+#include <ctype.h>
```
Comment (victorjulien) [inline src/detect-transform-urldecode.c:37]: this should be part of `suricata-common.h`.

File: src/detect-transform-urldecode.c [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+static void TransformUrlDecode(InspectionBuffer *buffer, void *options)
+{
+    const uint8_t *input = buffer->inspect;
+    const uint32_t input_len = buffer->inspect_len;
+    uint8_t output[input_len]; // we can only shrink
+
+    uint32_t output_size = BufferUrlDecode(input, input_len, output);
+
+    InspectionBufferCopy(buffer, output, output_size);
```
Comment (victorjulien) [inline src/detect-transform-urldecode.c:117]: can we avoid this step if the buffer is unchanged?

File: src/detect-transform-urldecode.h [hunk 5]
```diff
@@ -0,0 +1,30 @@
+/* Copyright (C) 2019 Open Information Security Foundation
```
Comment (victorjulien) [inline src/detect-transform-urldecode.h:1]: 2020

File: src/detect-http-client-body.c (hunk 2 above)
Comment (catenacyber) [reply src/detect-http-client-body.c:261]: ok

File: src/detect-transform-urldecode.c (hunk 1 above)
Comment (catenacyber) [reply src/detect-transform-urldecode.c:1]: Eheh, PR was submitted in 2019 ;-p

File: src/detect-transform-urldecode.h (hunk 5 above)
Comment (catenacyber) [reply src/detect-transform-urldecode.h:1]: ok

File: src/detect-transform-urldecode.c (hunk 3 above)
Comment (catenacyber) [reply src/detect-transform-urldecode.c:37]: indeed

File: src/detect-transform-urldecode.c (hunk 4 above)
Comment (catenacyber) [reply src/detect-transform-urldecode.c:117]: Ok.
This is the same in other transforms.
Should I create a optimization ticket in redmine ?

File: src/detect-http-client-body.c (hunk 2 above)
Comment (catenacyber) [reply src/detect-http-client-body.c:261]: S-V test no longer passes without this commit (and this PR does not make much sense either)
@jlucovsky what is next ?

**Pull request closed without merging**